include/zephyr/net/tftp.h

```diff
@@ -53,41 +44,120 @@ struct tftpc {
 #define TFTPC_REMOTE_ERROR       -4
 #define TFTPC_RETRIES_EXHAUSTED  -5
 
-/* @brief This function gets "file" from the remote server.
+/**
+ * @brief TFTP Asynchronous Events notified to the application from the module
+ *        through the callback registered by the application.
+ */
+enum tftp_evt_type {
+	/** DATA event when data is received from remote server.
+	 *
+	 * @note DATA event structure contains payload data and size.
+	 */
+	TFTP_EVT_DATA,
+
+	/** ERROR event when error is received from remote server.
+	 *
+	 * @note ERROR event structure contains error code and message.
+	 */
+	TFTP_EVT_ERROR
+};
+
+/** @brief Parameters for data event. */
+struct tftp_data_param {
+	uint8_t *data_ptr;         /**< Pointer to binary data. */
+	uint32_t len;              /**< Length of binary data. */
+};
+
+/** @brief Parameters for error event. */
+struct tftp_error_param {
+	char *msg;                 /**< Error message. */
+	int code;                  /**< Error code. */
+};
+
+/**
+ * @brief Defines event parameters notified along with asynchronous events
+ *        to the application.
+ */
+union tftp_evt_param {
+	/** Parameters accompanying TFTP_EVT_DATA event. */
+	struct tftp_data_param data;
+
+	/** Parameters accompanying TFTP_EVT_ERROR event. */
+	struct tftp_error_param error;
+};
+
+/** @brief Defines TFTP asynchronous event notified to the application. */
+struct tftp_evt {
+	/** Identifies the event. */
+	enum tftp_evt_type type;
+
+	/** Contains parameters (if any) accompanying the event. */
+	union tftp_evt_param param;
+};
+
+/**
+ * @typedef tftp_callback_t
  *
- * If the file is successfully received its size will be returned in
- * `client->user_buf_size` parameter.
+ * @brief TFTP event notification callback registered by the application.
+ *
+ * @param[in] evt Event description along with result and associated
+ *                parameters (if any).
+ */
+typedef void (*tftp_callback_t)(const struct tftp_evt *evt);
+
+/**
+ * @brief TFTP client definition to maintain information relevant to the
+ *        client.
+ */
+struct tftpc {
+	/** Pointer to the remote server address structure */
+	struct sockaddr *server;
```

Comment:
Storing only the pointer here can be a bit problematic. For example, the caller of this API might not realize that we will only store the pointer and allocate the server address from stack, in which case we might access already freed memory later. It would be better here if we copy the server address and not use a pointer.

Reply:
I had though about "struct sockaddr server" in the beginning, which requires some more bytes for the stucture.
Anyway application needs to config `server` or `*server` before using the APIs, so does `callback`. We don't have an API to COPY server contents yet. For simplicity, my proposal is to use `struct sockaddr server` and update API document requesting initialize of the client structure before API call.

Reply:
Hmm, I am sure that if we let the application developer to shoot him/herself to the foot, the developer will do that and we get the blame. IMHO, it would be better to design the API so that we are safe side and not let application design to crash the tftp library. This will consume a bit more memory but I think that is acceptable trade off here.
@rlubos / @mniestroj WDYT about this?

Reply:
I may be missing something here, but isn't the change you've requested @jukkar applied already? `server` is no longer a pointer.

Reply:
Proposed (and implemented) change looks good to me.

Reply:
Indeed, it was fixed already :+1: